cartofacade/cgo_api.go

```diff
@@ -0,0 +1,401 @@
+
+	if err := getErrorFromStatusCode(status); err != nil {
+		return err
+	}
+
+	status = C.viam_carto_add_sensor_reading_destroy(&resp.value)
+	if err := getErrorFromStatusCode(status); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// GetPosition is a wrapper for viam_carto_get_position
+func (vc *Carto) GetPosition(ctx context.Context) (GetPositionHolder, error) {
+	resp := GetPositionResponse{value: C.viam_carto_get_position_response{}}
+
+	status := C.viam_carto_get_position(vc.value, &resp.value)
+
+	if err := getErrorFromStatusCode(status); err != nil {
+		return GetPositionHolder{}, err
+	}
+
+	getPositionHolder := GetPositionHolder{
+		x: float64(resp.value.x),
+		y: float64(resp.value.y),
+		z: float64(resp.value.z),
+
+		ox:    float64(resp.value.o_x),
+		oy:    float64(resp.value.o_y),
+		oz:    float64(resp.value.o_z),
+		theta: float64(resp.value.theta),
+
+		real: float64(resp.value.real),
+		imag: float64(resp.value.imag),
+		jmag: float64(resp.value.jmag),
+		kmag: float64(resp.value.kmag),
+
+		compReference: bstringToGoString(resp.value.component_reference),
+	}
+
+	status = C.viam_carto_get_position_response_destroy(&resp.value)
+	if err := getErrorFromStatusCode(status); err != nil {
+		return GetPositionHolder{}, err
+	}
+
+	return getPositionHolder, nil
+}
+
+// GetPointCloudMapResponse is a struct to hold the c type viam_carto_get_point_cloud_map_response
+type GetPointCloudMapResponse struct {
+	value C.viam_carto_get_point_cloud_map_response
+}
+
+// GetPointCloudMap is a wrapper for viam_carto_get_point_cloud_map
+func (vc *Carto) GetPointCloudMap(ctx context.Context) ([]byte, error) {
+	// TODO: determine whether or not return needs to be a pointer for performance reasons
+	resp := GetPointCloudMapResponse{value: C.viam_carto_get_point_cloud_map_response{}}
+
+	status := C.viam_carto_get_point_cloud_map(vc.value, &resp.value)
+
+	if err := getErrorFromStatusCode(status); err != nil {
+		return nil, err
+	}
+
+	status = C.viam_carto_get_point_cloud_map_response_destroy(&resp.value)
+	if err := getErrorFromStatusCode(status); err != nil {
+		return nil, err
+	}
+
+	bytes := (*[]byte)(unsafe.Pointer(&resp.value.point_cloud_pcd))
+	return *bytes, nil
+}
+
+// GetInternalState is a wrapper for viam_carto_get_internal_state
+func (vc *Carto) GetInternalState(ctx context.Context) ([]byte, error) {
+	resp := GetInternalStateResponse{value: C.viam_carto_get_internal_state_response{}}
+
+	status := C.viam_carto_get_internal_state(vc.value, &resp.value)
+
+	if err := getErrorFromStatusCode(status); err != nil {
+		return nil, err
+	}
+
+	status = C.viam_carto_get_internal_state_response_destroy(&resp.value)
+	if err := getErrorFromStatusCode(status); err != nil {
+		return nil, err
+	}
+
+	bytes := (*[]byte)(unsafe.Pointer(&resp.value.internal_state))
+	return *bytes, nil
+}
+
+func bStringToGoStringSlice(bstr *C.bstring, length int) []string {
+	bStrings := (*[1 << 28]C.bstring)(unsafe.Pointer(bstr))[:length:length]
```

Comment:
Its not obvious to my why this works & whether or not it is correct. Could you please explain?

Reply:
Got it, very cool!